File: 394-decode-string/decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        nums,ret = 0,""
        for ch in s:
            if ch.isdigit():
                nums = nums*10 + int(ch)
            elif ch == "[":
                stack.append((nums,ret))
                nums,ret = 0,""
            elif ch == "]":
                old_nums,old_str = stack.pop(-1)
                ret = old_str + ret*old_nums
            else:
                ret +=ch
        return ret

File: 394-decode-string/decode-string.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<char> st;
        for(int i = 0; i< s.size(); i++){
            if(s[i] == ']'){
                string decode = "";
                while(st.top() != '['){
                    decode = st.top()+decode;
                    st.pop();
                }
                st.pop(); // -> '['
                int num = 0;
                int base = 1;
                while(!st.empty() && isdigit(st.top())){
                    num = num+(st.top()-'0')*base;
                    st.pop();
                    base*=10;
                }
                while(num!=0){
                    for(int j = 0; j<decode.size();j++){
                        st.push(decode[j]);
                    }
                    num --;
                }
            }
            else{
                st.push(s[i]);
            }
        }
        ///////////////
        string ret = "";
       while(!st.empty()){
            ret = st.top() + ret;
            st.pop();
       }
        
        return ret;
    }
};